lambda_src/lambda_function.py

```diff
@@ -97,15 +102,115 @@ def sync_flow(event: Any, context: Any = None) -> Dict[Text, Any]:
     LOG.debug('Destination header not found in a POST and hence using sync_flow().')
     headers = event['headers']
     req_body = loads(event['body'])
+    start_time = timer()
 
+    destination_driver = None
     batch_id = headers[BATCH_ID_HEADER]
     write_uri = headers.get('write-uri')
+
     LOG.debug(f'sync_flow() received destination: {write_uri}.')
 
     if write_uri:
         destination_driver = import_module(
             f'geff.drivers.destination_{urlparse(write_uri).scheme}'
         )
+
+    while True:
+        data = get_data_from_lock(batch_id)
+        if data is None:  # request hasn't been initialized
+            break
+        elif data != LOCKED:  # if false, the response is yet to be written
+            return data
+
+    open_lock(batch_id)  # initilaze the request
+
+    res_data = process_request(
+        req_body, headers, event, batch_id, write_uri, destination_driver
+    )
+
+    # Write data to s3 or return data synchronously
+    if write_uri:
+        response = destination_driver.finalize(  # type: ignore
+            write_uri, batch_id, res_data
+        )
+    else:
+        data_dumps = dumps({'data': res_data}, default=str)
+        response = {
+            'statusCode': 200,
+            'body': b64encode(compress(data_dumps.encode())).decode(),
+            'isBase64Encoded': True,
+            'headers': {'Content-Encoding': 'gzip'},
+        }
+        end_time = timer()
+        if response and (end_time - start_time) > 20:
+            LOG.debug('Storing the response in DynamoDB.')
+            try:
+                close_lock(batch_id, response)  # write the response
+            except ClientError as ce:
+                if ce.response['Error']['Code'] == 'ValidationException':
+                    LOG.error(ce)
+                    pass
+
+    if len(response) > 6_000_000:
+        response = response_size_error(response, req_body)
+    return response
+
+
+def response_size_error(
```

Comment:
Functions should be verbs describing what they do, so let's maybe call it `construct_size_error_response`? 

not sure if it should be part of this PR or a separate task, but if you're looking to refactor this piece, there appears to be a bug here that might be worth fixing in that `response` should probably consistently be a dict constructed by `create_response` with an appropriate type signature (i.e. keys `statusCode`, `body`, `isBase64Encoded`, `headers`)

Reply:
> Functions should be verbs describing what they do, so let's maybe call it construct_size_error_response?

Makes sense, will change it. 

> not sure if it should be part of this PR or a separate task, but if you're looking to refactor this piece, there appears to be a bug here that might be worth fixing in that response should probably consistently be a dict constructed by create_response with an appropriate type signature (i.e. keys statusCode, body, isBase64Encoded, headers)

It seems like `create_response` currently creates a dict with only `statusCode` and `body`, and the dict that does have `isBase64Encoded` and `headers`, is manually [created](https://github.com/Snowflake-Labs/geff/blob/2aa38df9a4345e40e53a6ee4dc261dae66be3398/lambda_src/lambda_function.py#L138). Would we want to maybe add `isBase64Encoded` and `headers` as optional args in `create_response` and use it?

(Also, after making the args optional, a replacement with the function [here](https://github.com/Snowflake-Labs/geff/blob/2aa38df9a4345e40e53a6ee4dc261dae66be3398/lambda_src/lambda_function.py#L85) and [here](https://github.com/Snowflake-Labs/geff/blob/2aa38df9a4345e40e53a6ee4dc261dae66be3398/lambda_src/lambda_function.py#L88) maybe?)